python/tests/plugins/validators/test__validate_used_definitions.py

```diff
@@ -0,0 +1,54 @@
+from aac.lang.definitions.collections import get_definition_by_name
+from aac.lang.definitions.lexeme import Lexeme
+from aac.lang.definitions.source_location import SourceLocation
+from aac.plugins.validators import ValidatorFindings, ValidatorResult
+from aac.plugins.validators.unused_definitions import validate_used_definitions
+
+from tests.active_context_test_case import ActiveContextTestCase
+from tests.helpers.context import get_core_spec_context
+from tests.helpers.parsed_definitions import create_schema_definition, create_field_entry
+
+
+class TestValidateUsedDefinitions(ActiveContextTestCase):
+    def test_validate_references_valid_definition_reference(self):
+        test_valid_definition_field = create_field_entry("ValidDefinitionField", "string")
+        test_definition = create_schema_definition("TestData", fields=[test_valid_definition_field])
+
+        expected_result = ValidatorResult()
+
+        test_active_context = get_active_context(reload_context=True)
+        test_active_context.add_definition_to_context(test_definition)
+        target_schema_definition = get_definition_by_name("Field", test_active_context.definitions)
+
+        actual_result = validate_used_definitions(test_definition, target_schema_definition, test_active_context)
+
+        self.assertEqual(expected_result.is_valid(), actual_result.is_valid())
+
+    def test_validate_references_invalid_definition_reference(self):
+        invalid_definition_type = "ThisTypeStringWontAppearInTheCoreSpecIHope"
+
+        test_invalid_definition_reference_field = create_field_entry("InvalidBehaviorField", invalid_definition_type)
+        test_invalid_schema_definition = create_schema_definition(
+            "InvalidSchema", fields=[test_invalid_definition_reference_field]
+        )
+
+        invalid_reference_error_message = ""
+        test_findings = ValidatorFindings()
+        expected_finding_location = SourceLocation(4, 10, 81, 42)
+        lexeme = Lexeme(expected_finding_location, "", "")
+        test_findings.add_error_finding(
+            test_invalid_schema_definition, invalid_reference_error_message, "validate thing", lexeme
+        )
+        expected_result = ValidatorResult([test_invalid_schema_definition], test_findings)
+
+        test_active_context = get_core_spec_context([test_invalid_schema_definition])
+        field_definition = test_active_context.get_definition_by_name("Field")
+
+        actual_result = validate_used_definitions(test_invalid_schema_definition, field_definition, test_active_context, "type")
+        actual_result_message = actual_result.get_messages_as_string()
+
+        self.assertEqual(expected_result.is_valid(), actual_result.is_valid())
+        self.assertEqual(expected_finding_location, actual_result.findings.get_error_findings()[0].location.location)
+        self.assertIn("Nonexistant", actual_result_message)
+        self.assertIn(invalid_definition_type, actual_result_message)
```

Comment:
I've also adjusted the test a bit to reflect the focus on finding references to our definition.
```suggestion
                test_unreferenced_schema_definition = create_schema_definition(
            "UnreferencedSchema", fields=[]
        )

        invalid_reference_error_message = ""
        test_findings = ValidatorFindings()
        expected_finding_location = SourceLocation(1, 8, 16, 18)
        lexeme = Lexeme(expected_finding_location, "", "")
        test_findings.add_info_finding(
            test_unreferenced_schema_definition, invalid_reference_error_message, "validate thing", lexeme
        )
        expected_result = ValidatorResult([test_unreferenced_schema_definition], test_findings)

        test_active_context = get_core_spec_context([test_unreferenced_schema_definition])
        field_definition = test_active_context.get_definition_by_name("Field")

        actual_result = validate_used_definitions(test_unreferenced_schema_definition, field_definition, test_active_context, "type")
        actual_result_message = actual_result.get_messages_as_string()

        self.assertEqual(expected_result.is_valid(), actual_result.is_valid())
        self.assertEqual(expected_finding_location, actual_result.findings.get_info_findings()[0].location.location)
        self.assertIn("not referenced", actual_result_message)
        self.assertIn(test_unreferenced_schema_definition.name, actual_result_message)

```
`invalid_definition_type` and `test_invalid_definition_reference_field` were removed since we just need to create a definition `test_unreferenced_schema_definition` that's not referenced by other definitions in order to prompt the unreferenced validation. Because of this, I've replaced a lot of the assertions previously referencing `invalid_definition_type` to `test_unreferenced_schema_definition` as we're just expecting findings indicating that `test_unreferenced_schema_definition` is not referenced by other definitions.

`expected_finding_location = SourceLocation(1, 8, 16, 18)` was updated since this is the lexeme position and span of `test_unreferenced_schema_definition`'s name `UnreferencedSchema` where the info finding will be presented to the user.

`self.assertEqual(expected_finding_location, actual_result.findings.get_error_findings()[0].location.location)` was updated to `self.assertEqual(expected_finding_location, actual_result.findings.get_info_findings()[0].location.location)` since we're expecting the linting validation to return an info level finding rather than and error level finding. 

Is also updated `self.assertIn("Nonexistant", actual_result_message)` to `self.assertIn("not referenced", actual_result_message)` as my suggested edit for the implementation changed the info message to reflect that the definition isn't referenced/used by other definitions. With the updated info message, I also updated the last assertion to `self.assertIn(test_unreferenced_schema_definition.name, actual_result_message)` to demonstrate that the definition that's not being referenced is being called out in the info message returned to the user.

Reply:
I need more information on this piece
```
expected_finding_location = SourceLocation(1, 8, 16, 18) was updated since this is the lexeme position and span of test_unreferenced_schema_definition's name UnreferencedSchema where the info finding will be presented to the user.
```
I wasn't confident about what I used initially at all, and I somewhat understand more the fields after checking the definition and your write up. But I am not at all following how you calculate/determine them?

Reply:
Ah, sorry. In the validator implementation we grab a lexeme (really it's a token) that's just some text in the AaC file. The line `reference_lexeme = definition_under_test.get_lexeme_with_value(definition_under_test.name)` is grabbing the line, column, position, and span of a token that we're targeting which is the definition's name in this case.
![Screenshot 2023-08-03 at 11 38 11 PM](https://github.com/jondavid-black/AaC/assets/60155541/f33ee5b8-2266-45f1-afb7-5f7bc12cf9b2)
In the screenshot below you can see the yaml equivalent of the definition we're using in our test. The definition's name is on line 2 (diagnostics are o indexed) and the definition's name `UnreferencedSchema` starts on the 8th character (0 indexed), is 18 characters long (the span), and lastly the word starts on the 16th character of the file (the position).
![Screenshot 2023-08-03 at 11 39 59 PM](https://github.com/jondavid-black/AaC/assets/60155541/3fdecb61-c447-49b3-b003-36139d8359f0)

Reply:
That helped make it a lot more clear, thank you!